clang/lib/Sema/SemaDecl.cpp

```diff
@@ -14690,6 +14690,13 @@ void Sema::FinalizeDeclaration(Decl *ThisDecl) {
     }
   }
 
+  // The result of __builtin_counted_by_ref cannot be assigned to a variable.
+  // It allows leaking and modification of bounds safety information.
+  if (IsBuiltinCountedByRef(VD->getInit()))
+    Diag(VD->getInit()->getExprLoc(),
+         diag::err_builtin_counted_by_ref_cannot_leak_reference)
+        << VD->getInit()->getSourceRange();
```

Comment:
Should we split this off into a helper function like `bool CheckInvalidBuiltinCountedByRef(const Expr *E);` ?

Reply:
I was going to comment on that too, but ended up forgetting about it; it’s used in like 3–4 places, so I personally would prefer making this a helper function.

Reply:
I'm not sure I agree. The diagnostic isn't the same in all places. And one of them takes arguments not available from the `Expr`. I would say that if we were adding several more diagnostics then a separate function would make sense. But for four, it's probably a bit more trouble than it's worth.